### ISSUE TITLE
GRPO Convergence Issue with vllm cuda graph enabled

### ISSUE THREAD

**Issue by ZhiyuLi-Nvidia** [NVIDIA-NeMo/RL#998]
**Describe the bug**

Found convergence collapse and spikes in logprob error with vllm cuda graph enabled

Repro script:
https://wandb.ai/nvidia/grpo-dev/runs/8ctnpv1m?nw=nwuserpchadha

**Steps/Code to reproduce bug**

Please list *minimal* steps or code snippet for us to be able to reproduce the bug.

A helpful guide on on how to craft a minimal bug report  http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports. 

**Expected behavior**

**Environment overview (please complete the following information)**

 - Environment location: [Bare-metal, Docker, Cloud(specify cloud provider - AWS, Azure, GCP, Collab)]
 - Method of install: [pip install or from source]. Please specify exact commands you used to install.
 - If method of install is [Docker], provide `docker pull` & `docker run` commands used

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- OS version
- PyTorch version
- Python version

**Additional context**

Example: GPU model

**Comment by parthchadha** [comment]
<img width="1182" height="762" alt="Image" src="https://github.com/user-attachments/assets/eab01ead-3f6b-4f55-88da-943a68a620b3" />

<img width="1157" height="761" alt="Image" src="https://github.com/user-attachments/assets/b1d3a250-2b23-4b5b-945f-160e4c324414" />

**Comment by ZhiyuLi-Nvidia** [comment]
@StrongerXi 

The accuracy improvement is not notable.

As you can see the exp with `deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B`, where we compare logprobs from policy(dtensor/mcore backend) and generation model(vllm backend):
* `baseline`: compilation with `use_inductor=True`
* `flag`: same compilation with `use_inductor=True` with addition flag `TORCHINDUCTOR_EMULATE_PRECISION_CASTS=1`
<img width="952" height="670" alt="Image" src="https://github.com/user-attachments/assets/c2c1111c-a9c9-427e-a8c8-bcd17bf77e49" />

All logprob error spikes disappear, If we either use eager mode or compliation with `use_inductor=False`.

**Comment by StrongerXi** [comment]
@ZhiyuLi-Nvidia thanks for checking. I have little context on this, but I did some search and found [this](https://docs.nvidia.com/nemo/rl/0.2.1/guides/grpo.html#multiplicative-token-probability-error-token-mult-prob-error). Sounds like `token_mult_prob_error` compares the log-probs between the inference model and the trained model.

I think this might be because the inference model (vllm uses torch.compile) and trained model (dtensor/mcore backend doesn't use torch.compile) have different numerics, so a tentative future solution is to enable inductor in vllm only when compile+inductor is enabled in the trained model as well. This way one should benefit from compile speed ups without accuracy loss.

Sanity check: I assume dtensor and mcore are separate backends? And what's the default & how does user config this?

**Comment by parthchadha** [comment]
> [@ZhiyuLi-Nvidia](https://github.com/ZhiyuLi-Nvidia) thanks for checking. I have little context on this, but I did some search and found [this](https://docs.nvidia.com/nemo/rl/0.2.1/guides/grpo.html#multiplicative-token-probability-error-token-mult-prob-error). Sounds like `token_mult_prob_error` compares the log-probs between the inference model and the trained model.
> 
> I think this might be because the inference model (vllm uses torch.compile) and trained model (dtensor/mcore backend doesn't use torch.compile) have different numerics, so a tentative future solution is to enable inductor in vllm only when compile+inductor is enabled in the trained model as well. This way one should benefit from compile speed ups without accuracy loss.
> 
> Sanity check: I assume dtensor and mcore are separate backends? And what's the default & how does user config this?

Hi @StrongerX, our logprob error metric is meant to exactly track what you mentioned, the error introduced in logprobs between training and generation frameworks. We have seen that this error is usually less than 5% for RL algos to converge. For all the other models we also use mcore with vllm (torch compile with triton kernels) and the convergence/logprob error looks fine.

**Comment by ZhiyuLi-Nvidia** [comment]
> @ZhiyuLi-Nvidia i'd like to reproduce this on the inductor end, and see where exactly the numerics fails, but your repro script is 404 for me. could you provide that in a different location?
@v0i0 hope this would be helpful to you
https://github.com/NVIDIA-NeMo/RL/blob/main/docs/adding-new-models.md#4vllm_precision_compilation_testpy

**Comment by qiaochuz-nv** [comment]
## QA RCCA Analysis

RCCA Status: Done
RCCA Category: covered
Test Action: none

Fix PR: https://github.com/NVIDIA-NeMo/RL/pull/1014
Root Cause: fix PR includes upstream test coverage.
Trigger Config: GRPO Convergence Issue with vllm cuda graph enabled

Coverage Assessment:
| Area | Result |
|---|---|
| Existing Test | Upstream fix PR test coverage |
| Coverage Status | COVERED |
| Gap | N/A |
| Corrective Action | No nmfw_tests addition required; fix PR includes upstream test coverage or equivalent regression coverage. |

Conclusion: RCCA classification completed for this closed NeMo-RL GitHub issue.
